File: dart/lib/src/wallet/wallet_manager.dart
/// Wallet Lifecycle and Account Management
///
/// Provides extended wallet management including:
/// - Wallet creation, restoration, and lifecycle
/// - Account and subaddress management
/// - Balance queries by account
library;

import 'dart:typed_data';

import 'wallet.dart';

/// Wallet type indicating available operations.
enum WalletType {
  /// Full wallet with spend key - can sign and send transactions
  full,

  /// View-only wallet - can see incoming transactions but cannot spend
  viewOnly,

  /// Offline wallet - has spend key but no network access, for cold signing
  offline,
}

/// Account information within a wallet.
class AccountInfo {
  /// Account index (0-based)
  final int index;

  /// Account label
  final String label;

  /// Primary address for this account
  final String primaryAddress;

  /// Total balance (confirmed + unconfirmed)
  final BigInt balance;

  /// Unlocked (spendable) balance
  final BigInt unlockedBalance;

  /// Number of subaddresses in this account
  final int subaddressCount;

  const AccountInfo({
    required this.index,
    required this.label,
    required this.primaryAddress,
    required this.balance,
    required this.unlockedBalance,
    required this.subaddressCount,
  });

  /// Create an empty account with default values.
  factory AccountInfo.empty(int index) => AccountInfo(
        index: index,
        label: index == 0 ? 'Primary account' : 'Account #$index',
        primaryAddress: '',
        balance: BigInt.zero,
        unlockedBalance: BigInt.zero,
        subaddressCount: 1,
      );
}

/// Result of wallet creation.
class WalletCreationResult {
  /// The created wallet instance
  final MoneroWallet wallet;

  /// 25-word mnemonic seed
  final List<String> mnemonic;

  /// Primary address of the wallet
  final String primaryAddress;

  const WalletCreationResult({
    required this.wallet,
    required this.mnemonic,
    required this.primaryAddress,
  });
}

/// Wallet-related exceptions.
sealed class WalletException implements Exception {
  final String message;
  const WalletException(this.message);

  @override
  String toString() => 'WalletException: $message';
}

class WalletNotFound extends WalletException {
  WalletNotFound(String path) : super('Wallet not found at: $path');
}

class InvalidPassword extends WalletException {
  const InvalidPassword() : super('Invalid wallet password');
}

class InvalidMnemonic extends WalletException {
  InvalidMnemonic(String reason) : super('Invalid mnemonic: $reason');
}

class ViewOnlyWalletError extends WalletException {
  ViewOnlyWalletError(String operation)
      : super('Cannot $operation with view-only wallet');
}

class OfflineWalletError extends WalletException {
  OfflineWalletError(String operation)
      : super('Cannot $operation with offline wallet');
}

class AccountNotFound extends WalletException {
  AccountNotFound(int index) : super('Account not found: $index');
}

class SubaddressNotFound extends WalletException {
  SubaddressNotFound(int account, int address)
      : super('Subaddress not found: $account/$address');
}

class StorageError extends WalletException {
  final Object? cause;
  StorageError(super.message, [this.cause]);

  @override
  String toString() =>
      cause != null ? 'StorageError: $message ($cause)' : super.toString();
}

/// Extended wallet interface with lifecycle management.
///
/// Provides wallet creation, restoration, and lifecycle operations.
abstract class WalletManager {
  // ─────────────────────────────────────────────────────────────────────────
  // Wallet Lifecycle
  // ─────────────────────────────────────────────────────────────────────────

  /// Create a new wallet with a fresh seed.
  ///
  /// Returns a [WalletCreationResult] containing the wallet and mnemonic.
  Future<WalletCreationResult> createWallet(WalletConfig config);

  /// Restore wallet from mnemonic seed.
  ///
  /// [mnemonic] should be a 25-word seed.
  /// [restoreHeight] is the block height to start scanning from.
  Future<MoneroWallet> restoreFromMnemonic({
    required WalletConfig config,
    required List<String> mnemonic,
    int restoreHeight = 0,
  });

  /// Restore wallet from keys (view-only or full).
  ///
  /// [spendKey] is null for view-only wallets.
  Future<MoneroWallet> restoreFromKeys({
    required WalletConfig config,
    required String address,
    required String viewKey,
    String? spendKey,
    int restoreHeight = 0,
  });

  /// Open an existing wallet file.
  ///
  /// Throws [WalletNotFound] if wallet doesn't exist.
  /// Throws [InvalidPassword] if password is wrong.
  Future<MoneroWallet> openWallet(WalletConfig config);

  /// Close wallet and release resources.
  ///
  /// Saves state before closing.
  Future<void> closeWallet(MoneroWallet wallet);

  /// Check if a wallet file exists at the given path.
  bool walletExists(String path);

  // ─────────────────────────────────────────────────────────────────────────
  // Wallet Info
  // ─────────────────────────────────────────────────────────────────────────

  /// Get wallet type (full, view-only, offline).
  WalletType getWalletType(MoneroWallet wallet);

  /// Get the mnemonic seed (only for full wallets).
  ///
  /// Throws [ViewOnlyWalletError] if wallet is view-only.
  Future<List<String>> getMnemonic(MoneroWallet wallet);

  /// Get wallet restore height.
  int getRestoreHeight(MoneroWallet wallet);

  /// Set wallet restore height (for re-scanning).
  Future<void> setRestoreHeight(MoneroWallet wallet, int height);
}

/// Account management operations.
abstract class AccountManager {
  /// Get all accounts in the wallet.
  Future<List<AccountInfo>> getAccounts(MoneroWallet wallet);

  /// Get a specific account.
  ///
  /// Throws [AccountNotFound] if account doesn't exist.
  Future<AccountInfo> getAccount(MoneroWallet wallet, int accountIndex);

  /// Create a new account.
  ///
  /// [label] is an optional label for the account.
  Future<AccountInfo> createAccount(MoneroWallet wallet, {String label = ''});

  /// Set account label.
  Future<void> setAccountLabel(
    MoneroWallet wallet,
    int accountIndex,
    String label,
  );

  /// Get balance for a specific account.
  Future<WalletBalance> getAccountBalance(MoneroWallet wallet, int accountIndex);

  /// Get total wallet balance (sum of all accounts).
  Future<WalletBalance> getTotalBalance(MoneroWallet wallet);
}

/// Subaddress management operations.
abstract class SubaddressManager {
  /// Get all subaddresses for an account.
  Future<List<SubaddressInfo>> getSubaddresses(
    MoneroWallet wallet,
    int accountIndex,
  );

  /// Create a new subaddress.
  ///
  /// [accountIndex] is the account to create subaddress in.
  /// [label] is an optional label.
  Future<SubaddressInfo> createSubaddress(
    MoneroWallet wallet,
    int accountIndex, {
    String label = '',
  });

  /// Set subaddress label.
  Future<void> setSubaddressLabel(
    MoneroWallet wallet,
    int accountIndex,
    int addressIndex,
    String label,
  );

  /// Get subaddress by indices.
  ///
  /// Throws [SubaddressNotFound] if subaddress doesn't exist.
  Future<SubaddressInfo> getSubaddress(
    MoneroWallet wallet,
    int accountIndex,
    int addressIndex,
  );

  /// Find which account/subaddress an address belongs to.
  ///
  /// Returns `(accountIndex, addressIndex)` or null if not found.
  Future<(int, int)?> findSubaddress(MoneroWallet wallet, String address);

  /// Get the number of subaddresses in an account.
  Future<int> getSubaddressCount(MoneroWallet wallet, int accountIndex);
}

/// Subaddress index pair for convenience.
class SubaddressIndex {
  final int major;
  final int minor;

  const SubaddressIndex({required this.major, required this.minor});

  /// Main account, primary address.
  static const main = SubaddressIndex(major: 0, minor: 0);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SubaddressIndex && major == other.major && minor == other.minor;

  @override
  int get hashCode => Object.hash(major, minor);

  @override
  String toString() => 'SubaddressIndex($major, $minor)';
}

/// Extended wallet balance with per-account breakdown.
class AccountBalance {
  final int accountIndex;
  final BigInt balance;
  final BigInt unlockedBalance;
  final BigInt pendingBalance;

  const AccountBalance({
    required this.accountIndex,
    required this.balance,
    required this.unlockedBalance,
    required this.pendingBalance,
  });

  static AccountBalance zero(int accountIndex) => AccountBalance(
        accountIndex: accountIndex,
        balance: BigInt.zero,
        unlockedBalance: BigInt.zero,
        pendingBalance: BigInt.zero,
      );
}

/// Output filter options for queries.
class OutputFilter {
  /// Include spent outputs
  final bool includeSpent;

  /// Include frozen outputs
  final bool includeFrozen;

  /// Filter by account index
  final int? accountIndex;

  /// Filter by subaddress index
  final int? subaddressIndex;

  /// Minimum amount
  final BigInt? minAmount;

  /// Maximum amount
  final BigInt? maxAmount;

  const OutputFilter({
    this.includeSpent = false,
    this.includeFrozen = true,
    this.accountIndex,
    this.subaddressIndex,
    this.minAmount,
    this.maxAmount,
  });

  static const all = OutputFilter(includeSpent: true, includeFrozen: true);
  static const spendable = OutputFilter(includeSpent: false, includeFrozen: false);
}

/// Transaction filter options for queries.
class TransactionFilter {
  /// Filter by account index
  final int? accountIndex;

  /// Filter by subaddress index
  final int? subaddressIndex;

  /// Include pending (unconfirmed) transactions
  final bool includePending;

  /// Only incoming transactions
  final bool? incoming;

  /// Minimum height
  final int? minHeight;

  /// Maximum height
  final int? maxHeight;

  /// Payment ID filter
  final String? paymentId;

  const TransactionFilter({
    this.accountIndex,
    this.subaddressIndex,
    this.includePending = true,
    this.incoming,
    this.minHeight,
    this.maxHeight,
    this.paymentId,
  });

  static const all = TransactionFilter();
  static const confirmed = TransactionFilter(includePending: false);
  static const incomingOnly = TransactionFilter(incoming: true);
  static const outgoingOnly = TransactionFilter(incoming: false);
}

File: kmp/monero-wallet/src/commonMain/kotlin/io/monero/wallet/WalletManager.kt
package io.monero.wallet

import java.math.BigInteger

/**
 * Wallet type indicating available operations.
 */
enum class WalletType {
    /** Full wallet with spend key - can sign and send transactions */
    FULL,
    /** View-only wallet - can see incoming transactions but cannot spend */
    VIEW_ONLY,
    /** Offline wallet - has spend key but no network access, for cold signing */
    OFFLINE
}

/**
 * Account information within a wallet.
 */
data class AccountInfo(
    /** Account index (0-based) */
    val index: Int,
    /** Account label */
    val label: String,
    /** Primary address for this account */
    val primaryAddress: String,
    /** Total balance (confirmed + unconfirmed) */
    val balance: BigInteger,
    /** Unlocked (spendable) balance */
    val unlockedBalance: BigInteger,
    /** Number of subaddresses in this account */
    val subaddressCount: Int
) {
    companion object {
        fun empty(index: Int) = AccountInfo(
            index = index,
            label = if (index == 0) "Primary account" else "Account #$index",
            primaryAddress = "",
            balance = BigInteger.ZERO,
            unlockedBalance = BigInteger.ZERO,
            subaddressCount = 1
        )
    }
}

/**
 * Extended wallet interface with account and lifecycle management.
 *
 * This extends the base [MoneroWallet] with:
 * - Account creation and management
 * - Subaddress labels
 * - Wallet type detection
 * - Restore height management
 */
interface WalletManager {

    // ─────────────────────────────────────────────────────────────────────────
    // Wallet Lifecycle
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Create a new wallet with a fresh seed.
     *
     * @param config Wallet configuration (path, password, network)
     * @return Created wallet with generated mnemonic
     */
    suspend fun createWallet(config: WalletConfig): WalletCreationResult

    /**
     * Restore wallet from mnemonic seed.
     *
     * @param config Wallet configuration
     * @param mnemonic 25-word mnemonic seed
     * @param restoreHeight Block height to start scanning from
     * @return Restored wallet
     */
    suspend fun restoreFromMnemonic(
        config: WalletConfig,
        mnemonic: List<String>,
        restoreHeight: Long = 0
    ): MoneroWallet

    /**
     * Restore wallet from keys (view-only or full).
     *
     * @param config Wallet configuration
     * @param address Primary address
     * @param viewKey Private view key (hex)
     * @param spendKey Private spend key (hex, null for view-only)
     * @param restoreHeight Block height to start scanning from
     * @return Restored wallet
     */
    suspend fun restoreFromKeys(
        config: WalletConfig,
        address: String,
        viewKey: String,
        spendKey: String? = null,
        restoreHeight: Long = 0
    ): MoneroWallet

    /**
     * Open an existing wallet file.
     *
     * @param config Wallet configuration with path and password
     * @return Opened wallet
     * @throws WalletException if wallet doesn't exist or password is wrong
     */
    suspend fun openWallet(config: WalletConfig): MoneroWallet

    /**
     * Close wallet and release resources.
     * Saves state before closing.
     */
    suspend fun closeWallet(wallet: MoneroWallet)

    /**
     * Check if a wallet file exists at the given path.
     */
    fun walletExists(path: String): Boolean

    // ─────────────────────────────────────────────────────────────────────────
    // Wallet Info
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Get wallet type (full, view-only, offline).
     */
    fun getWalletType(wallet: MoneroWallet): WalletType

    /**
     * Get the mnemonic seed (only for full wallets).
     *
     * @throws WalletException if wallet is view-only
     */
    suspend fun getMnemonic(wallet: MoneroWallet): List<String>

    /**
     * Get wallet restore height.
     */
    fun getRestoreHeight(wallet: MoneroWallet): Long

    /**
     * Set wallet restore height (for re-scanning).
     */
    suspend fun setRestoreHeight(wallet: MoneroWallet, height: Long)
}

/**
 * Account management operations.
 */
interface AccountManager {

    /**
     * Get all accounts in the wallet.
     */
    suspend fun getAccounts(wallet: MoneroWallet): List<AccountInfo>

    /**
     * Get a specific account.
     */
    suspend fun getAccount(wallet: MoneroWallet, accountIndex: Int): AccountInfo

    /**
     * Create a new account.
     *
     * @param label Optional label for the account
     * @return New account info
     */
    suspend fun createAccount(wallet: MoneroWallet, label: String = ""): AccountInfo

    /**
     * Set account label.
     */
    suspend fun setAccountLabel(wallet: MoneroWallet, accountIndex: Int, label: String)

    /**
     * Get balance for a specific account.
     */
    suspend fun getAccountBalance(wallet: MoneroWallet, accountIndex: Int): WalletBalance

    /**
     * Get total wallet balance (sum of all accounts).
     */
    suspend fun getTotalBalance(wallet: MoneroWallet): WalletBalance
}

/**
 * Subaddress management operations.
 */
interface SubaddressManager {

    /**
     * Get all subaddresses for an account.
     */
    suspend fun getSubaddresses(wallet: MoneroWallet, accountIndex: Int): List<SubaddressInfo>

    /**
     * Create a new subaddress.
     *
     * @param accountIndex Account to create subaddress in
     * @param label Optional label
     * @return New subaddress info
     */
    suspend fun createSubaddress(
        wallet: MoneroWallet,
        accountIndex: Int,
        label: String = ""
    ): SubaddressInfo

    /**
     * Set subaddress label.
     */
    suspend fun setSubaddressLabel(
        wallet: MoneroWallet,
        accountIndex: Int,
        addressIndex: Int,
        label: String
    )

    /**
     * Get subaddress by indices.
     */
    suspend fun getSubaddress(
        wallet: MoneroWallet,
        accountIndex: Int,
        addressIndex: Int
    ): SubaddressInfo

    /**
     * Find which account/subaddress an address belongs to.
     *
     * @return Pair of (accountIndex, addressIndex) or null if not found
     */
    suspend fun findSubaddress(wallet: MoneroWallet, address: String): Pair<Int, Int>?

    /**
     * Get the number of subaddresses in an account.
     */
    suspend fun getSubaddressCount(wallet: MoneroWallet, accountIndex: Int): Int
}

/**
 * Result of wallet creation.
 */
data class WalletCreationResult(
    val wallet: MoneroWallet,
    val mnemonic: List<String>,
    val primaryAddress: String
)

/**
 * Wallet-related exceptions.
 */
sealed class WalletException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class WalletNotFound(path: String) : WalletException("Wallet not found at: $path")
    class InvalidPassword : WalletException("Invalid wallet password")
    class InvalidMnemonic(reason: String) : WalletException("Invalid mnemonic: $reason")
    class ViewOnlyWallet(operation: String) : WalletException("Cannot $operation with view-only wallet")
    class OfflineWallet(operation: String) : WalletException("Cannot $operation with offline wallet")
    class AccountNotFound(index: Int) : WalletException("Account not found: $index")
    class SubaddressNotFound(account: Int, address: Int) : WalletException("Subaddress not found: $account/$address")
    class StorageError(message: String, cause: Throwable? = null) : WalletException(message, cause)
}
